### PR TITLE
feat: Make app surveys scalable

### PR DESCRIPTION
**Pull request by mattinannt**
Draft:
This PR includes the new API endpoint for user identification + survey filtering for Website surveys. Necessary changes in js-core are still to do.

## What does this PR do?

This PR adds user identification to website surveys. It allows Formbricks users using the `@formbricks/js/website` SDK to use `setUserId` to associate a survey response with a user and to synchronize the state of answered & viewed surveys across devices.

Fixes formbricks/internal#329

## How should this be tested?

- Use `formbricks.setUserId()` to set a userId in website surveys demo app.
- Check formbricks/js logs and local storage state if the changes have been synced.

### CODE REVIEW

**Review by mattinannt — COMMENTED** [review]
@pandeymangg Thank you for the updates 💪 I looked through the code; looks great 😊, I only found a few things that we can optimize or need to investigate:

File: packages/js-core/src/website/lib/widget.ts [hunk 1]
```diff
@@ -156,27 +136,55 @@ const renderWidget = async (
...
-        }
+
+        const responses = websiteConfig.get().personState.data.responses;
+        const newPersonState: TJsPersonState = {
+          ...websiteConfig.get().personState,
+          data: {
+            ...websiteConfig.get().personState.data,
+            responses: [...responses, surveyState.surveyId],
+          },
+        };
```
Comment (gupta-piyush19) [inline packages/js-core/src/website/lib/widget.ts:178]: Run this on the response creation only, as `onResponse` runs on each answer submission

Comment (gupta-piyush19) [reply packages/js-core/src/website/lib/widget.ts:178]: similar to `packages/js-core/src/app/lib/widget.ts#L197`

File: packages/js-core/src/shared/environmentState.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    data: { ...(state as TJsEnvironmentState["data"]) },
+    expiresAt: new Date(new Date().getTime() + 1000 * 60 * 30), // 30 minutes
+  };
+};
+
+export const addEnvironmentStateExpiryCheckListener = (
+  sdkType: "app" | "website",
+  config: AppConfig | WebsiteConfig
+): void => {
+  let updateInterval = 1000 * 30; // every 30 seconds
```
Comment (gupta-piyush19) [inline packages/js-core/src/shared/environmentState.ts:63]: Do we want to check the environment expiration after every 30 sec?

**Review by gupta-piyush19 — CHANGES_REQUESTED** [review]
Thanks, @pandeymangg, for the changes
The code changes look good. I have posted a few consistency-related comments. Can you please make changes accordingly?

File: apps/web/app/api/v1/client/[environmentId]/app/environment/lib/environmentState.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+ *
+ * @param environmentId
+ * @returns The environment state
+ * @throws ResourceNotFoundError if the environment or organization does not exist
+ * @throws InvalidInputError if the channel is not "app"
+ */
+export const getEnvironmentState = async (environmentId: string): Promise<TJsEnvironmentState["data"]> =>
+  cache(
+    async () => {
+      const environment = await getEnvironment(environmentId);
```
[the remark this answers is not in the document]
Comment (pandeymangg) [reply apps/web/app/api/v1/client/[environmentId]/app/environment/lib/environmentState.ts:34]: happens in the route before this 👍

File: packages/js-core/src/app/lib/person.ts [hunk 4]
```diff
@@ -15,11 +15,23 @@ export const logoutPerson = async (): Promise<void> => {
 export const resetPerson = async (): Promise<Result<void, NetworkError>> => {
   logger.debug("Resetting state & getting new state from backend");
   closeSurvey();
+
+  const userId = appConfig.get().personState.data.userId;
+  if (!userId) {
+    return err({
```
[the remark this answers is not in the document]
Comment (pandeymangg) [reply packages/js-core/src/app/lib/person.ts:21]: nope, we don't throw errors from the js-core package and return it to be logged properly from the global handler

File: packages/js-core/src/shared/utils.ts [hunk 5]
```diff
@@ -145,3 +145,87 @@ export const getDefaultLanguageCode = (survey: TSurvey) => {
...
+
+      default:
+        throw Error("Invalid displayOption");
+    }
+  });
+
+  // filter surveys that meet the recontactDays criteria
+  filteredSurveys = filteredSurveys.filter((survey) => {
+    // if no survey was displayed yet, show the survey
+    if (!lastDisplayAt) {
```
[the remark this answers is not in the document]
Comment (pandeymangg) [reply packages/js-core/src/shared/utils.ts:202]: thanks for pointing this out, fixed now 🙏

File: packages/js-core/src/shared/environmentState.ts (hunk 2 above)
Comment (pandeymangg) [reply packages/js-core/src/shared/environmentState.ts:63]: increased to a minute

**Comment by gupta-piyush19** [comment]
Thanks for the changes, @pandeymangg
LGTM 🚀

**Review by gupta-piyush19 — APPROVED** [review]

**Pull request merged**